kernel/aster-nix/src/time/timer.rs

```diff
@@ -0,0 +1,232 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{
+    boxed::Box,
+    collections::BinaryHeap,
+    sync::{Arc, Weak},
+    vec::Vec,
+};
+use core::time::Duration;
+
+use aster_frame::{arch::register_timer_interrupt_callback, sync::SpinLock};
+
+use super::{clock::Clock, timer_callback::TimerCallback, timespec_t, timeval_t};
+use crate::prelude::*;
+/// A timer with periodic functionality.
+///
+/// Setting the timer will trigger a callback function upon expiration of
+/// the set time. If the timer has a interval time set, it will use
+/// the interval time to configure a new timing after expiration.
+pub struct IntervalTimer {
+    interval: SpinLock<Duration>,
+    timer_manager: Arc<TimerManager>,
+    registered_callback: Box<dyn Fn() + Send + Sync>,
+    timer_callback: SpinLock<Weak<TimerCallback>>,
+}
+
+impl IntervalTimer {
+    /// Create a `IntervalTimer` instance from a [`TimerManager`].
+    /// The generated [`TimerCallback`] of this timer will be managed by the `TimerManager`.
+    fn new<F>(registered_callback: F, timer_manager: Arc<TimerManager>) -> Arc<Self>
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        Arc::new(Self {
+            interval: SpinLock::new(Duration::ZERO),
+            timer_manager,
+            registered_callback: Box::new(registered_callback),
+            timer_callback: SpinLock::new(Weak::default()),
+        })
+    }
+
+    /// Set the interval time for this timer.
+    /// The timer will be reset with the interval time upon expiration.
+    pub fn set_interval(&self, interval: Duration) {
+        *self.interval.lock_irq_disabled() = interval;
+    }
+
+    /// Clear the [`TimerCallback`] generated by this timer.
+    pub fn clear(&self) {
+        let timer_callback = self.timer_callback.lock_irq_disabled();
+        if let Some(timer_callback) = timer_callback.upgrade() {
+            timer_callback.cancel();
+        }
+    }
+
+    /// Set the timer with a timeout.
+    ///
+    /// It will generate a [`TimerCallback`], which will generally execute the callback
+    /// function when reaching timeout.
+    pub fn set_timer(self: &Arc<Self>, timeout: Duration, is_abs_time: bool) {
+        let expired_time = if is_abs_time {
+            timeout
+        } else {
+            let now = self.timer_manager.clock().read_time();
+            now + timeout
+        };
+        let timer_weak = Arc::downgrade(self);
+        let new_timer_callback = Arc::new(TimerCallback::new(
+            expired_time,
+            Box::new(move || interval_timer_callback(&timer_weak)),
+        ));
+
+        let mut timer_callback = self.timer_callback.lock_irq_disabled();
+        if let Some(timer_callback) = timer_callback.upgrade() {
+            timer_callback.cancel();
+        }
+        self.timer_manager.insert(new_timer_callback.clone());
+        *timer_callback = Arc::downgrade(&new_timer_callback);
```

Comment:
How do you like replacing the order of line 77 and line 78 to avoid cloning `new_timer_callback`?